WordPress/WordPressTest/ContextManagerMock+AutomaticTeardown.swift

```diff
@@ -0,0 +1,72 @@
+import XCTest
+
+@testable import WordPress
+
+extension ContextManagerMock {
+    /// Override `ContextManager.shared` with the mock instance, and un-override it when
+    /// given `testCase` finishes.
+    ///
+    /// This method should be called during test set up, typically from `XCTestCase` subclass's
+    /// `setUp` method.
+    ///
+    /// - Parameter testCase: The test case to wait for.
+    @objc func useAsSharedInstance(untilTestFinished testCase: XCTestCase) {
+        // Create the test observer singleton to add it to `XCTestObservationCenter`.
+        _ = AutomaticTeardownTestObserver.instance
+
+        let original = ContextManager.overrideInstance
+        ContextManager.internalSharedInstance()
+        ContextManager.overrideSharedInstance(self)
+
+        // This closure is going to be called by the test observer below when
+        // the test case finishes. The reason a combination of storing a closure in
+        // `XCTestCase` instance and a dedicated `XCTestObservation` implementation
+        // is used here is, we need to make sure tearing down the `ContextManager`
+        // mock instance happens _after_ test finishes execution.
+        //
+        // XCTest does have an API allowing us to add our own teardown block:
+        // `XCTestCase.addTeardownBlock`, but the added block is called _before_
+        // the test case's `tearDown` override method. Which means if this official
+        // API is used here instead, `ContextManager.shared` references two different
+        // objects during test execution: the mock instance before `tearDown`, or the
+        // real singleton during `tearDown`, which isn't ideal.
+        testCase.additionalTeardown = { [weak self] in
+            guard let self = self else {
+                return
+            }
+            self.mainContext.reset()
+            if ContextManager.overrideInstance === self {
+                ContextManager.overrideSharedInstance(original)
+            }
+        }
+    }
+}
+
+private class AutomaticTeardownTestObserver: NSObject, XCTestObservation {
+
+    static let instance = AutomaticTeardownTestObserver()
+
+    private override init() {
+        super.init()
+        XCTestObservationCenter.shared.addTestObserver(self)
+    }
+
+    func testCaseDidFinish(_ testCase: XCTestCase) {
+        testCase.additionalTeardown?()
+    }
+
+}
+
+private var additionalTeardownKey: Int = 0
```

Comment:
nit: can't this be a `let`? (given `objc_setAssociatedObject` expects an `UnsafeRawPointer`, not an `UnsafeMutableRawPointer`, I think a `let` would work?)

Reply:
Changing it to `let` produces an compiling error:
> error: cannot pass immutable value as inout argument: 'additionalTeardownKey' is a 'let' constant

But Swift compiler is happy if `withUnsafePointer` is used to get the address of the let variable. The result is slightly more verbose:
```swift
private let additionalTeardownKey: Int = 0
private extension XCTestCase {

    var additionalTeardown: (() -> Void)? {
        set {
            withUnsafePointer(to: additionalTeardownKey) { key in
                objc_setAssociatedObject(self, key, newValue, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN)
            }
        }
        get {
            withUnsafePointer(to: additionalTeardownKey) { key in
                objc_getAssociatedObject(self, key) as? (() -> Void)
            }
        }
    }

}

```

Reply:
Eh. Let's keep it as a `var` then. I'm surprised about the error because [this Apple doc](https://developer.apple.com/documentation/objectivec/1418509-objc_setassociatedobject) mentions the `key` being an `UnsafeRawPointer`, so not expecting a `…Mutable…` one… but 🤷